build_runner/lib/src/build_script_generate/build_script_generate.dart

```diff
@@ -21,23 +21,79 @@ const scriptSnapshotLocation = '$scriptLocation.snapshot';
 
 final _log = Logger('Entrypoint');
 
-Future<String> generateBuildScript() =>
-    logTimedAsync(_log, 'Generating build script', _generateBuildScript);
-
-Future<String> _generateBuildScript() async {
-  final builders = await _findBuilderApplications();
-  final library = Library((b) => b.body.addAll([
-        literalList(
-                builders,
-                refer('BuilderApplication',
-                    'package:build_runner_core/build_runner_core.dart'))
-            .assignFinal('_builders')
-            .statement,
-        _main()
-      ]));
-  final emitter = DartEmitter(Allocator.simplePrefixing());
-  try {
-    return DartFormatter().format('''
+/// Options to generate custom build scripts.
+///
+/// The build script is a generated Dart file in [scriptLocation]. It contains
+/// all builder definitions part of the build and is used by the top-level
+/// `build_runner` commands to run a build with the correct builders.
+///
+/// By providing custom generation options, it's possible to tweak parts of a
+/// build script.
+class BuildScriptGenerationOptions {
```

Comment:
This class feels a little clumsy to me. I'm not a fan of enumerating each bit of customization like this. I would prefer to build smaller blocks that can be put together, rather than a large block that allows fine-grained tweaks to small details.

My preference would be to do something like

```dart
Future<int> generateAndRun(
  List<String> args, {
  Logger logger,
  Future<String> Function() generateBuildScript = generateBuildScript,
  void Function(Object error, StackTrace stackTrace) handleUncaughtError,
}) async {
```

Then we can make `Future<Iterable<Expression>> _findBuilderApplications()` public and the consuming code could use that along with the APIs from `code_builder` to generate any type of build script they want.

We can also make the `PackageGraph`, and `Map<String, BuildConfig>` arguments to `findBuilderApplications` and let them be passed in rather than plumb through a `Future<PackageGraph> Function()` and `Future<Map<String, BuildConfig>> Function(PackageGraph, RunnerAssetReader)`.

This would impose a cost of some extra complexity to consumers - they will have to write the `main` themselves - but it gives maximal flexibility.

@jakemac53 WDYT?

Reply:
I don't have a strong opinion here - I did find the class a little bit surprising (more in the way that you implement it as opposed to passing it the options), but I was also OK with it.